lib/ecto/changeset.ex

```diff
@@ -1990,9 +1990,10 @@ defmodule Ecto.Changeset do
   """
   def check_constraint(changeset, field, opts \\ []) do
     constraint = opts[:name] || raise ArgumentError, "must supply the name of the constraint"
+    constraint_string = to_string(constraint)
     message    = message(opts, "is invalid")
     match_type = Keyword.get(opts, :match, :exact)
-    add_constraint(changeset, :check, to_string(constraint), match_type, field, {message, [constraint: :check]})
+    add_constraint(changeset, :check, constraint_string, match_type, field, {message, [constraint: :check, constraint_name: constraint_string]})
```

Comment:
given all constraint functions have a very similar shape now we could remove some duplication by calling `add_constraint` like this:

```elixir
add_constraint(changeset, :check, constraint_string, match_type, field, message)
```

in the case where constraint _type_ doesn't match what we want to put into metadata (e.g.: exclude&exclusion, no_assoc&foreign_key etc), we could do this:

```elixir
add_constraint(changeset, :exclude, constraint_string, match_type, field, message, :exclusion)
```